src/python/pants/backend/python/rules/hermetic_pex.py

```diff
@@ -1,34 +1,161 @@
 # Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
-
-from typing import Any, Iterable, Mapping, Optional
+import logging
+import os
+from dataclasses import dataclass
+from typing import Any, Iterable, Mapping, Optional, Tuple
 
 from pants.backend.python.subsystems.subprocess_environment import SubprocessEncodingEnvironment
+from pants.engine import process
+from pants.engine.engine_aware import EngineAware
 from pants.engine.fs import Digest
-from pants.engine.process import Process
+from pants.engine.process import BinaryPathRequest, BinaryPaths, Process
+from pants.engine.rules import collect_rules, rule
+from pants.engine.selectors import Get, MultiGet
 from pants.python.python_setup import PythonSetup
+from pants.subsystem.subsystem import Subsystem
+from pants.util.logging import LogLevel
+from pants.util.memo import memoized_property
+from pants.util.ordered_set import OrderedSet
 from pants.util.strutil import create_path_env_var
 
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class PexEnvironment(EngineAware):
+    path: Iterable[str]
+    interpreter_search_paths: Iterable[str]
+    bootstrap_python: Optional[str] = None
```

Comment:
Calculating this locally and then using it remotely is going to be problematic, because we don't know the contents of the remote PATH. But a solution would probably be to use the relative name of the interpreter so that it can continue to be looked up using the PATH.

Reply:
Oh, nevermind. On deeper inspection, I see that this attempts to implement PATH discovery via `BinaryPathRequest -> BinaryPaths`. I'm not sure that that will work at this level because the local runner defaults to a completely empty PATH (meaning it needs to be configured), while the remote runner defaults to the docker image's path. The strategy described in https://github.com/pantsbuild/pants/issues/9760#issuecomment-663208488 might be necessary to do this kind of discovery.

@Eric-Arellano : I would suggest backing out the portion of this that attempts to do PATH discovery in `@rules` for the above reasons, and let people hardcode the binary name (ie `python` vs `python3`) for now via an option.

Reply:
@stuhood I think you're missing what's going on here. First and foremost the PR fixed the associated issue and no more. As such, the primary driver is to separate PATH from the interpreter search path. In order. The easy step is using PEX_PYTHON_PATH as the interpreter search path leaving PATH for it's traditional use. I make no attempt to go beyond that status quo for setting PATH; i.e. <PATH> is a local expansion just like it has been for interpreter search paths. It's also a monolithic path that needs to have mixed local and remote values, just like the status qou.

Now, the key thing I think you missed: To bootstrap the PEX without pythons on the PATH, we need to say: `/full/path/to/interpreter my.pex ...`. The binary path lookup rules are used to do that and are handed the python interpreter search path to do that with. Hopefully makes sense now?

Reply:
> Now, the key thing I think you missed: To bootstrap the PEX without pythons on the PATH, we need to say: /full/path/to/interpreter my.pex .... The binary path lookup rules are used to do that and are handed the python interpreter search path to do that with. Hopefully makes sense now?

Yes, I think I believe that I understand what this is doing... but I think that it is broken for the reason I mentioned above. You cannot set the PATH from python and then search it from the context of a rule... that won't work with remoting unless you hardcode the PATH that you're expecting to exist in the remote host.

Reply:
Ok, thanks @jsirois : the path configuration is an independent issue that we can tackle via #9760, and the "binary discovery" within that PATH should be portable.

Resolving this thread then: non-issue!